Crash and race condition in etcd_meta_service

**Describe your problem**

We have noticed crash after pull request #313 merged, and yields the following backtrace:

```
(gdb) bt
#0  0x0000000000000008 in ?? ()
#1  0x00005650ecbfd386 in std::function<vineyard::Status (vineyard::Status const&, std::vector<vineyard::IMetaService::op_t, std::allocator<vineyard::IMetaService::op_t> > const&, unsigned int)>::function(std
::function<vineyard::Status (vineyard::Status const&, std::vector<vineyard::IMetaService::op_t, std::allocator<vineyard::IMetaService::op_t> > const&, unsigned int)> const&) (this=0x7f59be7fb648, __x=...)
    at /usr/include/c++/9/bits/std_function.h:263
#2  0x00005650ecbf7726 in std::pair<unsigned int, std::function<vineyard::Status (vineyard::Status const&, std::vector<vineyard::IMetaService::op_t, std::allocator<vineyard::IMetaService::op_t> > const&, unsi
gned int)> >::pair(std::pair<unsigned int, std::function<vineyard::Status (vineyard::Status const&, std::vector<vineyard::IMetaService::op_t, std::allocator<vineyard::IMetaService::op_t> > const&, unsigned in
t)> > const&) (this=0x7f59be7fb640) at /usr/include/c++/9/bits/stl_pair.h:303
#3  vineyard::EtcdWatchHandler::operator() (this=0x7f5990000b60, resp=...) at /home/gsbot/hetao/libvineyard/src/server/services/etcd_meta_service.cc:79
#4  0x00007f5a3b04049b in std::function<void (etcd::Response)>::operator()(etcd::Response) const (__args#0=..., this=0x7f59be7fba90) at /usr/include/c++/9/bits/std_function.h:683
#5  etcdv3::AsyncWatchAction::waitForResponse(std::function<void (etcd::Response)>) (this=this@entry=0x7f59b8004120, callback=...)
    at /home/gsbot/hetao/libvineyard/thirdparty/etcd-cpp-apiv3/src/v3/AsyncWatchAction.cpp:150
```

And it can also yield wrong result for the following sequence of requests:

```python
    client0 = vineyard.connect(vineyard_ipc_sockets[0])
    client1 = vineyard.connect(vineyard_ipc_sockets[1])
    for index in range(1024):
        o = client0.put('abcde...')
        client0.persist(o)
        client1.get_meta(o, sync_remote=True)
```

After many tirals there will be exception:

```python
        for index in range(1024):
            o = client0.put('abcde...')
            client0.persist(o)
>           client1.get_meta(o, sync_remote=True)
E           vineyard._C.ObjectNotExistsException: Object not exists: failed to read get_data reply: {"content":null,"type":"get_data_reply"}
```

--- a/src/server/services/etcd_meta_service.cc
+++ b/src/server/services/etcd_meta_service.cc
@@ -72,6 +72,13 @@ void EtcdWatchHandler::operator()(etcd::Response const& resp) {
     }
     }
   }
+  // Notes on [Execute order in boost asio context]
+  //
+  // The execution order is guaranteed to be the same with the post order.
+  //
+  // Ref: https://www.boost.org/doc/libs/1_75_0/doc/html/boost_asio/reference/
+  //      io_context__strand.html#boost_asio.reference.io_context__strand.orde
+  //      r_of_handler_invocation
   auto status = Status::EtcdError(resp.error_code(), resp.error_message());
   ctx_.post(boost::bind(callback_, status, ops, resp.index()));
   {

--- a/src/server/services/etcd_meta_service.cc
+++ b/src/server/services/etcd_meta_service.cc
@@ -16,6 +16,7 @@ limitations under the License.
 #include "server/services/etcd_meta_service.h"
 
 #include <chrono>
+#include <mutex>
 #include <string>
 #include <utility>
 #include <vector>
@@ -73,14 +74,18 @@ void EtcdWatchHandler::operator()(etcd::Response const& resp) {
   }
   auto status = Status::EtcdError(resp.error_code(), resp.error_message());
   ctx_.post(boost::bind(callback_, status, ops, resp.index()));
-  *handled_ = ((unsigned) resp.index());
-  // handle registered callbacks
-  while (!registered_callbacks_->empty()) {
-    auto iter = registered_callbacks_->front();
-    if (iter.first > (unsigned) resp.index())
-      break;
-    ctx_.post(boost::bind(iter.second, status, ops, resp.index()));
-    registered_callbacks_->pop();
+  {
+    std::lock_guard<std::mutex> scope_lock(registered_callbacks_mutex_);
+    handled_rev_.store(static_cast<unsigned>(resp.index()));
+
+    // handle registered callbacks
+    while (!registered_callbacks_.empty()) {
+      auto iter = registered_callbacks_.front();
+      if (iter.first > (unsigned) resp.index())
+        break;
+      ctx_.post(boost::bind(iter.second, status, ops, resp.index()));
+      registered_callbacks_.pop();
+    }
   }
 }
 
@@ -198,26 +203,28 @@ void EtcdMetaService::requestUpdates(
                       callback](pplx::task<etcd::Response> resp_task) {
     auto resp = resp_task.get();
     if (since_rev < (unsigned) resp.index() &&
-        since_rev + 1 > *this->handled_) {
+        since_rev + 1 > this->handled_rev_.load()) {
       if (this->watcher_) {
-        VLOG(10) << "------------------------- since: " << since_rev
+        VLOG(10) << "request updates: since: " << since_rev
                  << " current: " << resp.index()
-                 << " handled: " << *this->handled_;
-        this->registered_callbacks_->push(
-            std::pair<
-                unsigned,
-                callback_t<const std::vector<IMetaService::op_t>&, unsigned>>(
-                since_rev + 1, callback));
+                 << " handled: " << this->handled_rev_.load();
+        {
+          std::lock_guard<std::mutex> scope_lock(
+              this->registered_callbacks_mutex_);
+          this->registered_callbacks_.emplace(
+              std::make_pair(since_rev + 1, callback));
+        }
       } else {
         etcd_->watch(prefix_ + prefix, since_rev + 1, true)
-            .then(EtcdWatchHandler(server_ptr_->GetMetaContext(), callback,
-                                   prefix_, prefix_ + meta_sync_lock_,
-                                   this->registered_callbacks_,
-                                   this->handled_));
+            .then(EtcdWatchHandler(
+                server_ptr_->GetMetaContext(), callback, prefix_,
+                prefix_ + meta_sync_lock_, this->registered_callbacks_,
+                this->handled_rev_, this->registered_callbacks_mutex_));
       }
     } else {
-      server_ptr_->GetMetaContext().post(boost::bind(
-          callback, Status::OK(), std::vector<op_t>(), *this->handled_));
+      server_ptr_->GetMetaContext().post(
+          boost::bind(callback, Status::OK(), std::vector<op_t>(),
+                      this->handled_rev_.load()));
     }
   });
 }
@@ -230,7 +237,7 @@ void EtcdMetaService::startDaemonWatch(
         *etcd_, prefix_ + prefix, since_rev + 1,
         EtcdWatchHandler(server_ptr_->GetMetaContext(), callback, prefix_,
                          prefix_ + meta_sync_lock_, this->registered_callbacks_,
-                         this->handled_),
+                         this->handled_rev_, this->registered_callbacks_mutex_),
         true));
     this->watcher_->Wait([this, prefix, callback](bool cancalled) {
       if (cancalled) {

--- a/src/server/services/etcd_meta_service.h
+++ b/src/server/services/etcd_meta_service.h
@@ -17,6 +17,7 @@ limitations under the License.
 #define SRC_SERVER_SERVICES_ETCD_META_SERVICE_H_
 
 #include <memory>
+#include <mutex>
 #include <queue>
 #include <string>
 #include <utility>
@@ -46,14 +47,16 @@ class EtcdWatchHandler {
 #endif
       callback_t<const std::vector<IMetaService::op_t>&, unsigned> callback,
       std::string const& prefix, std::string const& filter_prefix,
-      std::shared_ptr<registered_callback_type> registered_callbacks,
-      std::shared_ptr<unsigned> handled)
+      registered_callback_type& registered_callbacks,
+      std::atomic<unsigned>& handled_rev,
+      std::mutex& registered_callbacks_mutex)
       : ctx_(ctx),
         callback_(callback),
         prefix_(prefix),
         filter_prefix_(filter_prefix),
         registered_callbacks_(registered_callbacks),
-        handled_(handled) {
+        handled_rev_(handled_rev),
+        registered_callbacks_mutex_(registered_callbacks_mutex) {
   }
 
   void operator()(pplx::task<etcd::Response> const& resp_task);
@@ -67,8 +70,10 @@ class EtcdWatchHandler {
 #endif
   const callback_t<const std::vector<IMetaService::op_t>&, unsigned> callback_;
   std::string const prefix_, filter_prefix_;
-  std::shared_ptr<registered_callback_type> const registered_callbacks_;
-  std::shared_ptr<unsigned> const handled_;
+
+  registered_callback_type& registered_callbacks_;
+  std::atomic<unsigned>& handled_rev_;
+  std::mutex& registered_callbacks_mutex_;
 };
 
 /**
@@ -115,8 +120,7 @@ class EtcdMetaService : public IMetaService {
       : IMetaService(server_ptr),
         etcd_spec_(server_ptr_->GetSpec()["metastore_spec"]),
         prefix_(etcd_spec_["prefix"].get_ref<std::string const&>()) {
-    this->registered_callbacks_ = std::make_shared<registered_callback_type>();
-    this->handled_ = std::make_shared<unsigned>(0);
+    this->handled_rev_.store(0);
   }
 
   void requestLock(
@@ -161,8 +165,10 @@ class EtcdMetaService : public IMetaService {
   std::shared_ptr<etcd::Watcher> watcher_;
   std::unique_ptr<asio::steady_timer> backoff_timer_;
   std::unique_ptr<boost::process::child> etcd_proc_;
-  std::shared_ptr<registered_callback_type> registered_callbacks_;
-  std::shared_ptr<unsigned> handled_;
+
+  registered_callback_type registered_callbacks_;
+  std::atomic<unsigned> handled_rev_;
+  std::mutex registered_callbacks_mutex_;
 
   friend class IMetaService;
 };

